### PR TITLE
Add address in user profile

### PR DESCRIPTION
**Pull request by Iamyuvraj143**
adding address feature
After login now user can add address in his profile

<kpd>address </kpd>

<kpd>address1 </kpd>

### CODE REVIEW

File: app/controllers/addresses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class AddressesController < ApplicationController
+  def create
+    @user = Current.user
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb:6]: Are you using this `@user` value somewhere?(except the next line). If not, do we really need to initialize this?

File: app/controllers/addresses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    @address = @user.addresses.create(address_params)
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: What if any validation fails?

File: app/controllers/addresses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    redirect_to user_path(@user)
+    flash[:notice] = "Address added Succesfully."
+  end
+
+  def destroy
+    @user = User.find(params[:user_id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Any use of this variable?

File: app/controllers/addresses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    @address = @user.addresses.find(params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: What's the difference between `find` and `find_by`?

Comment (kalpana19sharma) [reply app/controllers/addresses_controller.rb]: And what if the record is not found?

File: app/models/address.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Address < ApplicationRecord
+  belongs_to :user
```
Comment (kalpana19sharma) [inline app/models/address.rb:2]: Add some validations

File: app/views/addresses/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<h2>Add a address:</h2>
+<%= form_with model: [ @user, @user.addresses.build ] do |form| %>
```
Comment (kalpana19sharma) [inline app/views/addresses/_form.html.erb]: `@user.addresses.build` this logic should be in controller

File: app/views/welcome/_header.html.erb [hunk 7]
```diff
@@ -34,27 +31,26 @@
...
+            <li>  <%= link_to 'Profile', user_path(@user),  :class => 'dropdown-item'  %></li>
+            <li><hr class="dropdown-divider"></li>
+            <li>  <%= link_to 'Logout', logout_path,  method: :delete, :class => 'dropdown-item'  %></li>
+          </ul>
+        </li>
+           <% else %>
+            <li class="nav-item">
+              <%= link_to "Login", sign_in_path, :class => 'nav-link active'  %>
+            </li>
+             <% end %>
```
Comment (kalpana19sharma) [inline app/views/welcome/_header.html.erb]: Fix indentation

File: config/routes.rb [hunk 8]
```diff
@@ -5,5 +5,8 @@
   get 'sign_in', to: 'sessions#new'
   post 'sign_in', to: 'sessions#create', as: 'log_in'
   delete 'logout', to: 'sessions#destroy'
+  resources :users do
+    resources :addresses
+  end
```
Comment (kalpana19sharma) [inline config/routes.rb]: How many routes this logic will generate? And are we using all of the routes?

File: yarn.lock [hunk 9]
```diff
@@ -4807,6 +4807,11 @@ pnp-webpack-plugin@^1.7.0:
   dependencies:
     ts-pnp "^1.1.6"
 
+popper.js@^1.16.1:
+  version "1.16.1"
+  resolved "https://registry.yarnpkg.com/popper.js/-/popper.js-1.16.1.tgz#2a223cb3dc7b6213d740e40372be40de43e65b1b"
+  integrity sha512-Wb4p1J4zyFTbM+u6WuO4XstYx4Ky9Cewe4DWrel7B0w6VVICvPwdOpotjzcf6eD8TsckVnIMNONQyPIUFOUbCQ==
+
```
Comment (kalpana19sharma) [inline yarn.lock:4814]: What's the need of this?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 8 above)
Comment (Iamyuvraj143) [reply config/routes.rb]: its will generate all routes 
fixed by specifying only required routes

File: app/controllers/addresses_controller.rb (hunk 1 above)
Comment (Iamyuvraj143) [reply app/controllers/addresses_controller.rb:6]: No. variable removed

File: app/models/address.rb (hunk 5 above)
Comment (Iamyuvraj143) [reply app/models/address.rb:2]: Done

File: app/controllers/addresses_controller.rb (hunk 4 above)
Comment (Iamyuvraj143) [reply app/controllers/addresses_controller.rb]: If record not found exception handled

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -6,5 +6,15 @@ class ApplicationController < ActionController::Base
   def set_current_user
     Current.user = User.find_by(id: session[:user_id]) if session[:user_id]
   end
-  
+
+  def auth_check
+    # allows only logged in user
+    @user = User.find_by(id: session[:user_id])
+    if session[:user_id] != @user.id
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: Is it even possible? I mean first you are find a user with `session[:user_id]` and then matching if the ids are equal or not. This should always be false.

File: app/controllers/addresses_controller.rb [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+      redirect_to user_path(Current.user)
+      flash[:notice] = "Address added Succesfully."
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def edit
+    begin
+      @user = User.find_by(id: session[:user_id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: You are using this logic in multiple places. Can we extract this in a filter?

File: app/controllers/addresses_controller.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    rescue StandardError => e
+      redirect_to root_path, status: 303
+      flash[:notice] = "Something went Wrong.#{e}"
+    end
+  end
+
+   def update
+    begin
+      user = User.find_by(id: session[:user_id]) 
+      @address = user.addresses.find_by(id: params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: You are using this logic in multiple places. Can we extract this in a filter?

File: app/controllers/addresses_controller.rb [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      if @address!=nil
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: Use `.present?` or `.blank?`

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+class UsersController < ApplicationController
+  before_action :auth_check
+  
+  def show
+    @user = User.find_by(id: Current.user[:id])
```
Comment (kalpana19sharma) [inline app/controllers/users_controller.rb]: What's the need of this query? `Current.user` has same data, right?

File: app/views/addresses/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1 @@
+<%= render 'form' %>
```
Comment (kalpana19sharma) [inline app/views/addresses/edit.html.erb]: Newline

File: app/views/addresses/new.html.erb [hunk 16]
```diff
@@ -0,0 +1 @@
+<%= render 'form' %>
```
Comment (kalpana19sharma) [inline app/views/addresses/new.html.erb]: Newline

File: app/controllers/application_controller.rb [hunk 17]
```diff
@@ -6,5 +6,15 @@ class ApplicationController < ActionController::Base
...
+      redirect_to user_path(Current.user)
+      flash[:notice] = "Invalid opertion performed."
+    end
+  end
+
 end
+
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: I think there are some whitespaces here.

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -11,11 +11,10 @@
...
 
   <body>
-    <%= render 'welcome/header' %>
+    <%=render 'welcome/header' %>
     <% if notice || alert %>
-      <p class="notice"><%= notice %></p>
-      <p class="alert"><%= alert %></p>
-  <% end %>
-  <%= yield %>
+      <%= render 'welcome/alert' %>
```
Comment (kalpana19sharma) [inline app/views/layouts/application.html.erb]: You can save these partials in `layouts` folder. `welcome` folder is dedicated to only welcome page content.

File: config/routes.rb [hunk 19]
```diff
@@ -5,5 +5,8 @@
   get 'sign_in', to: 'sessions#new'
   post 'sign_in', to: 'sessions#create', as: 'log_in'
   delete 'logout', to: 'sessions#destroy'
+  resources :users, only: [:show] do
+    resources :addresses, only: [:new, :edit, :create, :update, :destroy]
+  end
```
Comment (kalpana19sharma) [inline config/routes.rb]: Use `%i` syntax for array of symbols.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/addresses_controller.rb [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def edit
+    begin
+    rescue StandardError => e
+      redirect_to root_path, status: 303
+      flash[:notice] = "Something went Wrong => .#{e}"
+    end
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: When we have no code in begin rescue block then don't use it.

File: app/controllers/addresses_controller.rb [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+          redirect_to user_path(@user)
+           flash[:notice] = "Address updated succesfully."
+        else
+          render :edit, status: :unprocessable_entity
+        end
+      end
+    rescue StandardError => e
+      redirect_to root_path, status: 303
+      flash[:notice] = "Something went Wrong.#{e}"
+    end
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: If you are going to use this logic in each action then you should find a better and optimized way to do this.

File: app/controllers/addresses_controller.rb [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+  private
+
+  def address_params
+    params.require(:address).permit(:address_line, :city, :state, :zipcode, :country)
+  end
+
+  def load_user
+    @user = User.find_by(id: session[:user_id]) 
+    @address = @user.addresses.find_by(id: params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]: If the address is not found then user should get a proper error message.

File: app/controllers/application_controller.rb [hunk 23]
```diff
@@ -6,5 +6,14 @@ class ApplicationController < ActionController::Base
   def set_current_user
     Current.user = User.find_by(id: session[:user_id]) if session[:user_id]
   end
-  
+
+  def auth_check
+    # allows only logged in user
+    @user = User.find_by(id: session[:user_id])
+    if session[:user_id] != @user.id
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: What's the need of this?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/addresses_controller.rb [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+  private
+
+  def address_params
+    params.require(:address).permit(:address_line, :city, :state, :zipcode, :country)
+  end
+
+  def load_user
+      @user = Current.user
+      @address = @user.addresses.find_by(id: params[:id])
+      if !@address.present?
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]:
```suggestion
      unless @address.present?
```

File: app/controllers/addresses_controller.rb [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+        redirect_to user_path(@user)
+        flash[:notice] = "Something wnet Wrong :- Address does not exist."
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]:
```suggestion
        flash[:notice] = "Something went Wrong :- Address does not exist."
```

File: app/controllers/addresses_controller.rb [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+      end
+  end
+
+  def auth_check
+    # allows only logged in user
+    @user = User.find_by(id: params[:user_id])
+    if @user.nil? || Current.user.id != @user.id
```
Comment (kalpana19sharma) [inline app/controllers/addresses_controller.rb]:
```suggestion
    if Current.user.id != @user&.id
```

File: app/controllers/application_controller.rb [hunk 27]
```diff
@@ -6,5 +6,14 @@ class ApplicationController < ActionController::Base
...
-  
+
+  def auth_check
+    # allows only logged in user
+   @user = User.find_by(id: params[:id])  
+    if @user.nil? || Current.user.id != @user.id  
+      redirect_to root_path
+      flash[:notice] = "Invalid opertion performed. "
+    end
+  end
```
Comment (kalpana19sharma) [inline app/controllers/application_controller.rb]: What's the need of this now? Maybe you can shift this message to users controller if needed.

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**